### DISCUSSION TITLE
Action to generate Global params

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#19563] [Q&A General] [answered]
Hello,
So currently i'm putting together an action to initialise a lot of variables, auxVariables, and ultimately form global params with the names of these variables. I'm aware of the global params action, but I am not sure how to add it to my action.

so far what i'm doing is adding the variables when the _current_task = "add_aux_variable" and _current_task = "add_variable", similar to PolycrystalVariablesAction.

however, I'm not sure how to set the global parameters in the custom action.  

here's the code I'm using:

```
registerMooseAction("CustomApp", CustomVariablesAction, "set_global_params");

void
CustomVariablesAction::act()
{
 
  if (_current_task = "set_global_params")
  {
    auto globalparams = _factory.getValidParams("GlobalParamsAction");
    globalparams.addParam<std::string>("rho_vars") = _rho_vars;
    globalparams.addParam<std::string>("rho_virt_vars") = _rho_virt_vars;
  }
}
```

Thank you in advance.

**Comment by vanwdani** [comment, marked as answer]
Well, there are a couple of ways to do this.  The `GlobalParamAction` needs to know what the the expected `inputParameter` will be on the left side of your conditional statement and the given name on the right.  A typical example would be something like
`[GlobalParams]
    temperature = temperature
  []`
This works well for manually creating input files, but not as well for action creation.  Adding the parameters like you have shown is usually done within the `InputParameters` section of the `Action`.  Once these parameters have been defined, then they are available for the user to enter within the corresponding `Action` block.  This would be something like:
`[MyAction]
    rho_vars = "whatever_you_enter"
  []`
So for this action since you are already creating the `Variables`, and `AuxVariables`, and the `InitialConditions` if they have any, you already know what the names will be.  I think you should be able to pass these names directly to whatever class needs them, be it a `Material`, `BC`, `PostProcessor`, or another `Action`.  If this isn't the case and you want the names from the `GlobalParams` block to define the names created and then passed, `cli_args` is probably the easiest route.
https://mooseframework.inl.gov/application_usage/input_syntax.html

**Reply by KhaledNabilSharafeldin** [reply]
I do have the names generated by the code, and stored in one string space separated similar to the one passed in the input file

```
std::string _rho_vars = "var1a var1b var1c var2a "
```

what I just understood was that this:
```
    auto globalparams = _factory.getValidParams("GlobalParamsAction");
    globalparams.addParam<std::string>("rho_vars") = _rho_vars;
```

only CREATES the rho_vars but doesnt fill in the values, is that correct?

also, I'm populating this string `_rho_vars` during `add_variable` and `add_aux_variable`, so does `set_global_params` get called before or after these time flags?

thank you for your reply!

**Reply by vanwdani** [reply]
```
auto globalparams = _factory.getValidParams("GlobalParamsAction");
globalparams.addParam<std::string>("rho_vars") = _rho_vars;
```

The 1st line is getting the already defined `inputParameters` from the `GlobalParamsAction`, which looking at the file is basically on the the `inputParameters` of `Action.C`.  The 2nd line is not placed in the right area.  This should be done within a `InputParameters` block within the class before initialization.  What this does is says, "this class is expecting the user to give me the names of the rho variables."  You've already created those names within the action, so you can't directly input them.

So the question is:  Where are these parameters going to be applied? 

 If you already know where these need to go, you can create the needed classes within your action.  Just like you've used `add_variable`, you can utilize `add_material` or `add_bc`, etc.  Then it's easily done with something like 

```
auto type = NAME_OF_CLASS;
auto params = _factory.getValidParams(type);
params.set<TYPE_OF_INPUT_PARAMETER_EXPECTED_BY_CLASS>("NAME_WITHIN_CLASS) = _rho_vars;
_problem->addMaterial(type, NAME_OF_BLOCK_YOU_CHOOSE, params);
```
